### DISCUSSION TITLE
Shiny `Glass Card` from Logseq landing page

### DISCUSSION THREAD

**Discussion by stdword** [stdword/logseq13-full-house-plugin#9] [Showroom]
###### Inspired by reddit [post](https://www.reddit.com/r/logseq/comments/11ya189/how_do_i_replicate_what_is_on_the_logseqcom_front/) in `r/logseq`

<p align="center">
    <b>Let's replicate the book card from Logseq.com main page 💪</b>
    <br/>
    <br/>
    <img src="https://user-images.githubusercontent.com/1984175/229943053-82aea6fb-3cd1-497a-967c-d76533a6cff4.png" width=45% />
    <img src="https://user-images.githubusercontent.com/1984175/229949828-01222090-bdbc-4a52-b39e-30dc5f91e2ce.gif" width=45% />
    <details closed><summary align="center">video</summary>
        <video width="40%" src="https://user-images.githubusercontent.com/1984175/229942344-816a3053-f935-4446-a55e-984eea41d17f.mp4"/>
    </details>
</p>

## `🏛view` for pages
1) Create _glass-card_ `🏛view` (use copy button in the top right corner):
<details closed><summary>code</summary>

````markdown
- template:: glass-card
  template-version:: 11
  plugin-version:: 4.0.0
  link:: [Shiny `Glass Card` from Logseq landing page](https://github.com/stdword/logseq13-full-house-plugin/discussions/9)
  template-list-as:: view
  template-usage:: `:page [[{|}]]`
  arg-cover::
  arg-title:: .name
  arg-info1:: @author
  arg-info2:: @category
  arg-info3::
  arg-card-width::
  arg-cover-height::
  arg-cover-width:: 56px
    - ```jsx
      <style>
        #``c.identity.slot`` .glass-card {
          background-image: linear-gradient(to right bottom, var(--gradient-from), 55%, var(--gradient-to));
          --gradient-from: color-mix(in srgb, var(--ls-selection-text-color) 35%, transparent);
          --gradient-to: color-mix(in srgb, var(--color-level-6) 10%, transparent);
          border-radius: 0.5rem;
          backdrop-filter: blur(100px);
        }
      
        #``c.identity.slot`` .glass-card > div {
          background-color: color-mix(in srgb, var(--ls-primary-background-color) 50%, transparent);
          border-radius: 0.5rem;
          margin: 0.5px;
          padding: 15px 0.75rem 0.75rem;
          ``when(c.args['card-width'], 'min-width: $1;')``
          ``when(c.args['card-width'], 'max-width: $1;')``
        }
      
        #``c.identity.slot`` .glass-card > div > div {
          /* compatibility with Bullet Threading plugin */
          height: auto !important;
        }
      
        #``c.identity.slot`` .glass-card img {
          ``when(c.args['cover-width'], 'width: $1; min-width: $1;')``
          ``when(c.args['cover-height'], 'height: $1; min-height: $1;')``
        }
      
        #``c.identity.slot`` .glass-card div > a > strong {
          color: color-mix(in srgb, var(--ls-link-text-hover-color) 80%, transparent);
        }
      
        #``c.identity.slot`` .glass-card .info {
          font-size: 0.9rem;
          font-weight: 300;
          line-height: 1.5rem;
        }
      </style>
      
      ``{_
          var mode = c.args.block ? 'block' : 'page'
          var reference = mode === 'block' ? c.block.uuid : c.page.name_
          
          var title = when(c.args.title, dev.get(mode + c.args.title) || c.args.title)
          var info1 = when(c.args.info1, dev.get(mode + c.args.info1) || c.args.info1)
          var info2 = when(c.args.info2, dev.get(mode + c.args.info2) || c.args.info2)
          var info3 = when(c.args.info3, dev.get(mode + c.args.info3) || c.args.info3)
          
          if (title) {
              if (Array.isArray(title))
                  title = title[0]
              if (title.startsWith('[[') && title.endsWith(']]'))
                  title = title.slice(2, -2)
          }
          if (info1) {
              info1 = Array.isArray(info1) ? info1 : [info1]
            info1 = dev.toHTML(info1.join('\n'))
          }
          if (info2) {
              info2 = Array.isArray(info2) ? info2 : [info2]
            info2 = dev.toHTML(info2.join('\n'))
          }
         if (info3) {
              info3 = Array.isArray(info3) ? info3 : [info3]
            info3 = dev.toHTML(info3.join('\n'))
          }
      
          var markup
          if (!c.args.cover && mode === 'page') {
              const tree = top.logseq.api.get_page_blocks_tree(c.page.name)
              markup = tree[0].children[0].content
              if (!markup)
                markup = tree[1].content
          }
          else
              markup = when(c.args.cover, dev.get(mode + c.args.cover) || c.args.cover)
      
            var cover = ''
          if (markup) {
            [ cover ] = dev.links(markup)
            if (!cover)
                    cover = dev.asset(markup)
          }
      _}``
      
      <div class="glass-card">
          <div class="flex">
              <div class="flex items-center">
                  <a data-on-click="clickRef" data-ref="``reference``">
                        ``{_ if (cover) {   _}``
                        <img src="``cover``"
                        ``_ when(c.args['cover-width'], 'width="$1"') _``
                        ``_ when(c.args['cover-height'], 'height="$1"') _``
                        />``{ }   _}``
                  </a>
              </div>
              ``{_ if (title || info1 || info2 || info3) {   _}``
                  <div class="info flex-col px-3 opacity-80">
                      ``{_ if (title) {  _}``
                          <a data-on-click="clickRef" data-ref="``reference``">
                              <strong class="text-2xl font-medium">``title``</strong></a>``{ }   _}``
                          ``_ when(info1, '<div class="pt-2 pb-1">$1</div>') _``
                          ``_ when(info2, '<div class="pt-1 pb-1">$1</div>') _``
                          ``_ when(info3, '<div class="pt-2"     >$1</div>') _``
                  </div>``{ }   _}``
        </div>
      </div>
      ```
````
</details>

2) Create «Intertwingled» book page. There are properties which _glass-card_ relies on: `name`, `author`, `category` and the <ins>link to book cover</ins> from **the second** page block (you could change it: see Tuning section).
   ```markdown
   - author:: [[Peter Morville]]
     name:: Intertwingled
     category:: [[📖/philosophy]]
     - ![cover](../assets/book-cover_intertwingled.jpg)
   ```
3) Note: You need to setup [Awesome Links](https://github.com/yoyurec/logseq-awesome-links) plugin in order to display icon near the author's page reference. Install plugin and add `icon` property to `[[Peter Morville]]` page: `icon:: 👤`.
4) Add rendering `🏛view` code to some block and enjoy book card:
   ```clojure
   {{renderer :template-view, glass-card, :page [[Intertwingled]] }}
   ```

## `🏛view` for blocks
0) Example provided for following block structure:
    ````markdown
    Irriducibile
    cover:: ![](https://www.ibs.it/images/9788804739036_0_536_0_75.jpg){:height 76, :width 39}
    title:: Irriducibile
    subtitle:: La coscienza, la vita. i computer e la nostra natura
    author:: [[Federico Faggin]]
    area:: Philosophy
    status:: [[Reading]]
    year:: 2022
    ````
1) Create the same `🏛view` as for pages
2) Add _inherited_ `🏛view` for blocks:
   ````markdown
    - template:: glass-card for block
      template-version:: 3
      plugin-version:: 4.0.0
      link:: [Shiny `Glass Card` from Logseq landing page](https://github.com/stdword/logseq13-full-house-plugin/discussions/9)
      template-usage:: ``:block ({|})``
      template-list-as:: view
      arg-cover:: @cover
      arg-title:: @title
      arg-info1:: @author
      arg-info2:: @area
      arg-info3::
      arg-card-width::
      arg-cover-height::
      arg-cover-width:: 56px
        - ``await layout('glass-card', layout.args('block', 'cover', 'title', 'info1', 'info2', 'info3', 'card-width', 'cover-height', 'cover-width', 'title-size'))``
   ````
3) Add rendering `🏛view` code to some block and enjoy book card. **Note**: use `:block` argument as content source for the view:
   ```clojure
   {{renderer :template-view, book-card, :block ((643338ba-…)) }}
   ```

## Tuning the `🏛view`
<img width="531px" src="https://user-images.githubusercontent.com/1984175/230956563-08bda7dd-0f15-4974-816b-c4fbee8186f0.png"/>
<img width="333px" src="https://user-images.githubusercontent.com/1984175/230956941-76a344dd-583e-43c1-88e3-f1253cd48e97.png"/>

- You could change the displaying content with `:cover`, `:title`, `:info1`, `:info2` & `:info3` arguments (or with its corresponding `arg-`-prefixed properties in template block):
  ```clojure
  {{renderer :template-view, book-card, :block ((643338ba-…)), :title @subtitle, :info3 @year, :card-width 470px, :cover-width 100px}}
  {{renderer :template-view, book-card, :block ((643338ba-…)) }}
  {{renderer :template-view, book-card, :block ((643338ba-…)), :cover ""}}
  {{renderer :template-view, book-card, :block ((643338ba-…)), :title "", :info1 "", :info2 ""}}
  ```
  - `@year` means `🏛view` would access `year` property of provided page or block
  - Special case for `cover` argument and _page_ source: if no property would be provided the value would be retrieved from **the second** page block
- Use `:title-size` to change title's size. Available choices can be found [here](https://tailwindcss.com/docs/font-size).
  - `{{renderer :template-view, book-card, :page [[Intertwingled]], :title-size text-xl}}`
- Use`:card-width` to change card's size (it is automatic by default):
  - `{{renderer :template-view, book-card, :page [[Intertwingled]], :card-width 350px}}`
- Use `:cover-height` & `:cover-width`: to change card's cover size
  - `{{renderer :template-view, book-card, :page [[Intertwingled]], :cover-width 56px, :cover-height 84px}}`

- ⚠️ The card style is adapted only for _Dark Themes_. Left a vote [here](https://github.com/stdword/logseq13-full-house-plugin/issues/10), if you're looking for _Light Themes_ support.

## Additional info
- Logseq v0.9.20
- Default Dark Theme
- Plugins: [Banners](https://github.com/yoyurec/logseq-banners-plugin) & [Awesome Links](https://github.com/yoyurec/logseq-awesome-links)

**Comment by Gitbreast** [comment]
wow, this is cool, out of my imagination. Is it possible to adapt itself to the light/dark mode?
The plugin looks like, such an almighty content inserter XD

**Reply by stdword** [reply]
🙏 You are so right! This is the idea itself: bring JS ability to write scripts based on our Logseq data to program any behavior. It is a bit slower than Logseq clojure core, but more agile and powerful 💪

**Reply by stdword** [reply]
You can adapt colors to any theme - they are brought out to properties as params (so you don’t need to understand the code itself). But it is still a ui design work you need to do.
We can try to build universal card colors (to mimic to theme), but it may loose that sexy look.

**Reply by Gitbreast** [reply]
@stdword I just simply followed the steps then got a little unusual like this, any idea?
<img width="300" alt="image" src="https://user-images.githubusercontent.com/96460796/230178917-98a6033c-8ed3-4793-ba44-b3000f03643b.png">

**Reply by stdword** [reply]
> @stdword I just simply followed the steps then got a little unusual like this, any idea? 

I guess it was a bug while I copy the code from Logseq. Fixed now — check it, please 🙏

But I don't know about the cover alignment. Does it misaligned in other themes?

**Reply by Gitbreast** [reply]
@stdword I disabled all awesome plugins and css then pick the default theme, it still looks like this. Compared with your gif, it looks like not the cover’s problem, it’s the rest elements alignment, moved down a little, the cover’s position is right. 
btw, I choose a png and use a jpg name, do they have anything to do with each other?
<img width="300" alt="image" src="https://user-images.githubusercontent.com/96460796/230279368-b22fe06a-ef1d-4a97-9aeb-83e1d2cfa9a6.png">

Update: I found out it’s the compatibility issue with `Bullet Threading` plugin.  https://github.com/pengx17/logseq-plugin-bullet-threading
<img width="300" alt="image" src="https://user-images.githubusercontent.com/96460796/230420161-ccfedcf0-404a-425b-bed6-801838ccc621.png">

**Reply by stdword** [reply]
Thank you! I'll check it.

And you may check the new version of code: it fixes the shift of book card. Github Markup makes a big pain while inserting Logseq code blocks as foldable block of github code =(

**Reply by stdword** [reply]
@Gitbreast fixed compatibility with Bullet Threading plugin 🙌

**Reply by Gitbreast** [reply]
@stdword wow, it works all right now. I was even trying to minimize the card size, but I couldn’t find the font size value of book title, which one it is? I just started learning web develop, hope one day I could understand all the code XD
<img width="325" alt="image" src="https://user-images.githubusercontent.com/96460796/231385248-62b6bf50-eaa9-40f8-adc3-c60bbf64b87e.png">

Plus, I feel like adding icon to every author page isn’t the best way, is it possible to fetch the property icon setting of [Awesome Props](https://github.com/yoyurec/logseq-awesome-props)?
> Install plugin and add icon property to [[Peter Morville]] page: icon:: 👤.

**Reply by stdword** [reply]
Hi!

1️⃣
I've just added:
<img width="600px" src="https://user-images.githubusercontent.com/1984175/231409860-9aa0246d-38aa-4d67-9737-fdbfed3ff950.png"/>
Choices: https://tailwindcss.com/docs/font-size#class-reference

2️⃣
> Plus, I feel like adding icon to every author page isn’t the best way, is it possible to fetch the property icon setting of [Awesome Props](https://github.com/yoyurec/logseq-awesome-props)?

Could you provide more info on how you want to use Awesome Props here?
In guide I've mention [Awesome Links](https://github.com/yoyurec/logseq-awesome-links), not Props. Instead of setting icon for every author's page you can set icon for `[[Author]]` page and then use `type:: [[Author]]` (but it still requires you to set `type` for every author)

**Reply by Gitbreast** [reply]
@stdword 1. I can’t believe I miss that part XD. But it’s not work after adding `:title-size text-*`.
render as below: `{{renderer :template-view, glass-card, :page [[Intertwingled-test]], :title-size text-xs}}`
<img width="220" alt="image" src="https://user-images.githubusercontent.com/96460796/231459667-a2fe7c7b-047b-457e-8213-60243702c46a.png">
2. A book page has its properties, and properties have their "Awesome Props icon”. At first I thought could use these icons to replace the manually setting icon in [[Author]], to show on the book card. But now I consider if the linear style of tables-icon(which is Awesome-Props using) fits the glass style of book card. Just a fleeting thought, maybe delay it, what do you think?
<img width="235" alt="image" src="https://user-images.githubusercontent.com/96460796/231462396-80b832e2-371f-491b-a9ec-1c94ef02f923.png">

**Reply by stdword** [reply]
@Gitbreast, no-no I just added param `:title-size` (it is new one) — you need to update view body (copy glass-card code again)

**Reply by stdword** [reply]
About properties: I get it! The problem is that Awesome Links plugin is responsible for displaying an icon near author's name (but not Full House). So, I literally cant handle this in Full House.

But I guess you can use the same icon with Tabler Icons plugin... Oh, no. I've checked: it doesn't work for now. I guess we need to wait for @yoyurec to fix it: Awesome Links cant display icons from Tabler Icons now.
<img src="https://user-images.githubusercontent.com/1984175/231473911-c985f62d-9b81-4723-8509-0d622db4c051.png" width="150px"/>

But you still can use icons from Nerd font (some of them). eg for `nf-fa-user_circle`:
<img src="https://user-images.githubusercontent.com/1984175/231476652-cc2baa3b-8ebe-42b5-a9a3-8114accd1707.png" width="400px"/>

**Reply by Gitbreast** [reply]
> About properties: I get it! The problem is that Awesome Links plugin is responsible for displaying an icon near author's name (but not Full House). So, I literally cant handle this in Full House.

@stdword The point of my thought was, is there a way to automatically get the prop icon of property “author” and show it on the card, before the value of it. But as you say, handling the displaying would be another big project and it’s not worth it. So just ignore it.

I played a while in `custom.css`, tried to write a little CSS to override the adaptive color of my using light/dark theme. 
Now it looks better. The renderer can even be used multiple times in a markdown table of a single block, maybe could use it as a note bookshelf.
<img width="400" alt="image" src="https://user-images.githubusercontent.com/96460796/231542771-f119b24a-5f68-4ea4-aa06-17eb62a73495.png"> <img width="400" alt="image" src="https://user-images.githubusercontent.com/96460796/231542911-a9b1a3de-9b90-4dc0-977a-7aab86a9d2bb.png">
I just paste my CSS code here as a copy and for others reference.
```css
/* book card adaptive */
.dark-theme #main-content-container .glass-card {
  background-image: linear-gradient(to bottom right, #3e3f45 70%, #232427 99%);
}
.dark-theme #main-content-container .glass-card .flex{
  background-color: #20212480;
}
.dark-theme #main-content-container .glass-card div > a > strong {
  color: #cccccc ;
}

.light-theme.white-theme #main-content-container .glass-card {
  background-image: linear-gradient(to bottom right, #dfdedc 50%, #f4f4f4 99%);
}
.light-theme.white-theme #main-content-container .glass-card .flex{
  background-color: #ffffff80 ;
}
.light-theme.white-theme #main-content-container .glass-card div > a > strong {
  color: #433f38 ;
}
```

**Reply by stdword** [reply]
Wow! Looks very cool 💪

**Idea**: use kanban board with columns (To Read, Reading, Done) and move book cards between them 😎 🙌

I've created issues to improve this tuning card UX. Left the links here if you want to subscribe to it:
- https://github.com/stdword/logseq13-full-house-plugin/issues/10
- https://github.com/stdword/logseq13-full-house-plugin/issues/11

**Reply by Gitbreast** [reply]
@stdword Indeed! `#.kanban` has its pros.
<img width="873" alt="image" src="https://user-images.githubusercontent.com/96460796/231704400-240ae107-734c-4727-b872-989cd874c8c9.png">
It can even display in journals query, funny! Maybe later I would try to write a specific “reading book” query for it.
<img width="1201" alt="image" src="https://user-images.githubusercontent.com/96460796/231703833-6224b129-c0c2-480f-87d2-18aa96f0538d.png">

**Comment by alxlg** [comment]
Would it be possible to do the same with a block as book and the cover as block-property value?

**Reply by stdword** [reply]
Absolutely yes. There is no restrictions: we can construct card (and any other view) from data structure we need.

I can provide an example of how to do that later, if you need this exact card view, but from block. Or you just wondering if that is possible?

**Reply by alxlg** [reply]
@stdword I generally save books as blocks with a property like `cover:: ![](cover.png)` so that it appears in table view of queries. So I just need to play with your plugin? If you can show how it would be helpful, thanks!

**Reply by hewcaw** [reply]
+1, I have a similar use case too!

**Reply by stdword** [reply]
@alxlg @hewcaw , guys, I've started work on this. Could you share a couple of your book blocks examples, please?

**Reply by alxlg** [reply]
@stdword I hope this helps:

```
Irriducibile 
Class:: [[book]]
Title:: Irriducibile
Subtitle:: La coscienza, la vita. i computer e la nostra natura
Author:: [[Federico Faggin]]
Area:: [[Philosophy]]
Status:: [[Reading]]
Cover:: ![](https://www.ibs.it/images/9788804739036_0_536_0_75.jpg)
Year:: 2022
```

**Reply by stdword** [reply]
@alxlg @hewcaw Hi! I've updated the guide: take a look at the **Tuning the `🏛view`** section

And update `🏛 Full House Templates` to latest version (2.4.0)

**Reply by Gitbreast** [reply]
Got some time to tune a little bit to fit my own request. Thanks to your work it works well with a simple intuitive tuning.👍 @stdword 
1. About title: I use **page name as book title**, so directly using `c.page.name` is perfect.
```js
    var title = ''
    if (c.args.title && mode === 'block') {
      	title = dev.get(mode + c.args.title) || c.args.title }
    else if (mode === 'page') {
      	title = c.page.name
    }
    else { title = 'No Title' }
```
2. About cover: A book card would looks not good without a cover, but sometimes I just want to leave it there or delay it, so I tune a bit and add a default cover for it. I **give up the option of using second block as cover, always using the template property** `arg-cover:: @cover` . Now, a book card would **show the default cover if there’s no cover property for the book (block&page mode)**. A little question: I don’t know whether the`[ ]` around `cover` is necessary in this situation?

```js
    var markup
    /* if (!c.args.cover)
		markup = top.logseq.api.get_page_blocks_tree(c.page.name)[1].content
      	markup = c.args.info3
    else */
  	markup = when(c.args.cover, dev.get(mode + c.args.cover))

   	var cover = ''
    /* change the path of default cover below */
    if (!markup) {
      	cover = dev.links('![universal_cover.png](../assets/universal_cover_1690401349961_0.png)')
        if (!cover) {
      		cover = dev.asset('![universal_cover.png](../assets/universal_cover_1690401349961_0.png)')
        }
    }
    else if (markup) {
    	[ cover ] = dev.links(markup)
    	if (!cover) {
      		cover = dev.asset(markup)
        }
    }
```
without cover property: <img width="200" alt="image" src="https://github.com/stdword/logseq13-full-house-plugin/assets/96460796/6d15e9f0-86b6-46fa-9cf1-f01e6021a96e"> with cover property: <img width="200" alt="image" src="https://github.com/stdword/logseq13-full-house-plugin/assets/96460796/4b536b1d-7124-4ecb-9b06-0fdf2c3ede99">

3. More: I try to use book card in page property and use query function to read it. But it failed, seems like Logseq’s not support it. 
And, is it a way to custom the open area of the book page, open in main area instead of the right side?
Update: Have no idea why it open in right side back then, now it indeed opens in middle, but could it support Ctrl-click open in a new Tabs? Seems like need to develop with tabs, another long story I suppose. XD

**Reply by stdword** [reply]
Hi! Nice! Thank you for kind words 🙏

1) In order to use page name in page mode you can specify arg `title` as `.name` (no «@», but «.»), without code changes.

2) > I don’t know whether the[ ] around cover is necessary in this situation?

   Yes, it is necessary, as of `dev.links` always return an Array.

   > dev.asset('![universal_cover.png] (../assets/universal_cover_1690401349961_0.png)')

   For `dev.asset` you can use only asset name, without markup and prefix `../assets/`:
   ```typescript
   dev.asset('universal_cover_1690401349961_0.png')
   ```

   And I've updated template and now you can specify default asset name in template property `arg-cover`.

3) Yes, it is not supported in Logseq. Glass Card is only a view — Logseq internally see it as string `{{renderer :view ...}}`

**Comment by stdword** [comment]
Just updated the view: it should nicely display glass card for **any** DARK theme

**Comment by JorgeFadul** [comment]
Hello, I'm new at Logseq and I don't know how to properly add a view, I have Full House extension, I copy the code glass-card view, and I copy the example and I get this:
![Logseq error](https://github.com/stdword/logseq13-full-house-plugin/assets/54223010/1aa4ae5e-8ec4-459e-89f2-708aee0917ce)


And a blank space
Can someone help me with an explanation please? I'll be really grateful.

**Reply by stdword** [reply]
Hello! Yep, there was a problem. Just fixed it: copy template code again.
Thank you for a feedback  🙏

**Reply by JorgeFadul** [reply]
Thank you, it works perfectly

**Comment by duiliuliu** [comment]
Hello, I've designed a new card template, but I've found that the rendering result does not meet my expectations.

```jsx
<style>
   
 #``c.identity.slot`` body {
    font-family: 'Arial', sans-serif;
    background-color: #f0f0f0;
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100vh;
    margin: 0;
}

 #``c.identity.slot`` .ticket {
    position: relative;
    width: 100%;
    max-width: 520px;
    border-radius: 10px;
    padding: 20px;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
    overflow: hidden;
}

 #``c.identity.slot`` .ticket-number {
    font-size: 24px;
    font-weight: bold;
    color: red;
}

 #``c.identity.slot`` .station {
    display: flex;
    justify-content: space-between;
    align-items: center;
    font-size: 20px;
    margin: 10px 0;
}

 #``c.identity.slot`` .arrow {
    margin: 0 10px;
    font-size: 24px;
}

 #``c.identity.slot`` .date-train {
    display: grid;
    grid-template-columns: 1fr 1fr;
    gap: 10px;
    margin: 10px 0;
}

 #``c.identity.slot`` .additional-info {
    font-size: 14px;
    margin: 10px 0;
}

 #``c.identity.slot`` .bottom-section {
    display: flex;
    justify-content: space-between;
    border-top: 1px solid #ccc;
    padding-top: 10px;
    margin-top: 10px;
}

 #``c.identity.slot`` .qr-code {
    width: 46px;
    height: 46px;
}

 #``c.identity.slot`` .barcode {
    margin-top: 10px;
    height: 24px;
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=');
    background-repeat: repeat-x;
}

</style>

``{_
    var mode = c.args.block ? 'block' : 'page'
     
    var title = when(c.args.title, dev.get(mode + c.args.title) || c.args.title)
    var info1 = when(c.args.info1, dev.get(mode + c.args.info1) || c.args.info1)
    var info2 = when(c.args.info2, dev.get(mode + c.args.info2) || c.args.info2)
    var info3 = when(c.args.info3, dev.get(mode + c.args.info3) || c.args.info3)
    var markup = when(c.args.cover, dev.get(mode + c.args.cover) || c.args.cover)
    var ref_link = when(c.args.ref_link, dev.get(mode + c.args.ref_link) || c.args.ref_link)

    
       function createElement(tag, attributes, ...children) {
            const element = document.createElement(tag);
            for (const key in attributes) {
                if (key.startsWith('on')) {
                    element.addEventListener(key.substring(2).toLowerCase(), attributes[key]);
                } else {
                    element.setAttribute(key, attributes[key]);
                }
            }
            children.forEach(child => {
                if (typeof child === 'string') {
                    element.appendChild(document.createTextNode(child));
                } else {
                    element.appendChild(child);
                }
            });
            return element;
        }

        function generateTicket({
            fromStation = 'Beijing',
            toStation = 'SH',
            trainNumber = '01-002',
            date = '2024-04-05',
            departureTime = '08:28',
            color = 'green'
        }) {
            const colorMap = {
                green: '#e8f6f6',
                blue: '#e8f0f6',
                pink: '#f6e8e8',
                yellow: '#f6f3e8',
            };

            const id = Math.random().toString(36).substring(2, 10).toUpperCase() + 'Q';

            const ticketDiv = createElement('div', {
                class: 'ticket',
                style: `background-color: ${colorMap[color]};`
            });

            // Ticket Number
            const ticketNumberDiv = createElement('div', { class: 'ticket-number' }, id);
            ticketDiv.appendChild(ticketNumberDiv);

            // Stations
            const stationsDiv = createElement('div', { class: 'station' },
                createElement('span', null, `${fromStation} 站`),
                createElement('span', { class: 'arrow' }, '→'),
                createElement('span', null, `${toStation} 站`)
            );
            ticketDiv.appendChild(stationsDiv);

            // Date and Train Info
            const dateTrainDiv = createElement('div', { class: 'date-train' },
                createElement('div', null, `${new Date(date).toLocaleDateString('zh-CN')} ${departureTime}开`),
                createElement('div', null, `${trainNumber}号`)
            );
            ticketDiv.appendChild(dateTrainDiv);

            // Additional Info
            const additionalInfoDiv = createElement('div', { class: 'additional-info' },
                createElement('div', null, '新至通特等座'),
                createElement('div', null, '限乘当日当次车')
            );
            ticketDiv.appendChild(additionalInfoDiv);

            // Bottom Section with QR Code
            const bottomDiv = createElement('div', { class: 'bottom-section' });

            const infoDiv = createElement('div', { class: 'info' },
                createElement('div', { class: 'mb-1' },
                    createElement('span', null, '买票请到12306'),
                    createElement('span', null, '发货请到95306')
                ),
                createElement('div', { class: 'insurance' }, '中国铁路旅客运输保险')
            );
            bottomDiv.appendChild(infoDiv);

            // QR Code
            const qrCodeDiv = createElement('img', {
                src: `https://api.qrserver.com/v1/create-qr-code/?data=ticket:${id}&size=46x46`,
                class: 'qr-code'
            });
            bottomDiv.appendChild(qrCodeDiv);

            ticketDiv.appendChild(bottomDiv);

            // Barcode placeholder
            const barcodeDiv = createElement('div', { class: 'barcode' });
            ticketDiv.appendChild(barcodeDiv);

            return ticketDiv;
        }

   	var ticket_card = generateTicket({
            fromStation: '北京',
            toStation: '上海',
            trainNumber: 'G101',
            date: '2024-05-01',
            departureTime: '14:30',
            color: 'pink'
        })
_}``

<div> 
      ``_ `<div>${ticket_card}</div>` _``
    </div>
</div>
</div>
```

block:
{{renderer :template-view, ticket-card, :block  , :online}}

block render result:
<img width="284" alt="image" src="https://github.com/user-attachments/assets/c7e4571b-fc81-4d08-984b-96a0fcd0126c">

**Reply by duiliuliu** [reply]
my expected result ：
<img width="516" alt="image" src="https://github.com/user-attachments/assets/24c1f60a-1549-42e6-9b45-2257e2d7f31a">

**Reply by stdword** [reply]
The issue doesn't related to the plugin, but in Javascript. You cannot just create HTMLDevElement and convert it to string via .toString() method. Use .outerHTML instead:

```javascript
...
            // Barcode placeholder
            const barcodeDiv = createElement('div', { class: 'barcode' });
            ticketDiv.appendChild(barcodeDiv);

            return ticketDiv.outerHTML;
        }
...
```

**Comment by N4gasak1** [comment]
Hello, I've tried the showcase but the render command give me an error back. "Cannot read properties of undefined (reading 'content')" in line 40 of the jsx copy-pastet content.
I'm not really familiar.... would does it means?

**Reply by stdword** [reply]
Hello! Sorry for the late reply.
Just checked the code — there is no errors. But the issue is in the page data: the cover should be the child block of the first page block. Just fixed the example. Try it again

**Reply by N4gasak1** [reply]
Thanks man! But probably it didn't worked.
This ist the error I get, when I insert your `{{renderer :template-view, glass-card, :page [[<pagename>]] }}`
<img width="420" height="310" alt="grafik" src="https://github.com/user-attachments/assets/8befc642-edf8-40d2-95e2-9e9ff6237b6a" />
"Cannot read properties of undefined (reading 'content')"

The error is still there and I thought it would be in the code :thinking:.

**Reply by N4gasak1** [reply]
Never mind, I figured it out. But yet I don't understand how to adapt the example :sweat_smile: 
Thanks man :pray:

**Reply by stdword** [reply]
@N4gasak1 but it works. No error in the code itself. But it may be related to the context of the execution. Here is the demo:

https://github.com/user-attachments/assets/edb0b213-b276-4e51-a1d8-32787569cae6
